subworkflows/local/params_check.nf

```diff
@@ -45,26 +46,66 @@ workflow PARAMS_CHECK {
                 outdir: outdir,
             ],
             file_fasta,
+            file(fasta + ".fai"),
         ]
     }
 
-    // Only flow to gunzip when required
+    // Identify the Fasta files that are compressed
     ch_parsed_fasta_name = ch_input_files.branch {
-        meta, filename ->
-            compressed : filename.getExtension().equals('gz')
+        meta, fasta, fai ->
+            compressed : fasta.getExtension().equals('gz')
             uncompressed : true
     }
 
-    // gunzip .fa.gz files
-    ch_unzipped_fasta   = GUNZIP ( ch_parsed_fasta_name.compressed ).gunzip
+    // uncompress them, with some channel manipulations to maintain the triplet (meta, fasta_fai)
+    ch_unzipped_fasta   = GUNZIP ( ch_parsed_fasta_name.compressed.map { meta, fasta_gz, fai -> [meta, fasta_gz] } ).gunzip
+                            .join(ch_parsed_fasta_name.compressed).map { meta, fasta, fasta_gz, fai -> [meta, fasta, fai] }
     ch_versions         = ch_versions.mix(GUNZIP.out.versions.first())
 
-    // Combine with preexisting .fa files
-    ch_plain_fasta      = ch_parsed_fasta_name.uncompressed.mix(ch_unzipped_fasta)
+    // Check if the faidx index is present
+    ch_parsed_fasta_name.uncompressed.mix(ch_unzipped_fasta).branch {
+        meta, fasta, fai ->
+            indexed : fai.exists()
+            notindexed : true
+    } . set { ch_inputs_checked }
 
+    // Generate Samtools index and chromosome sizes file, again with some channel manipulations
+    ch_samtools_faidx   = CUSTOM_GETCHROMSIZES (ch_inputs_checked.notindexed.map { meta, fasta, missing_fai -> [meta, fasta] }).fai
+                            .join(ch_inputs_checked.notindexed).map { meta, fai, fasta, missing_fai -> [meta, fasta, fai] }
+    ch_versions         = ch_versions.mix(CUSTOM_GETCHROMSIZES.out.versions)
+
+    // Read the .fai file, extract sequence statistics, and make an extended meta map
+    ch_fasta_fai = ch_inputs_checked.indexed.mix(ch_samtools_faidx).map {
+        meta, fasta, fai -> [meta + get_sequence_map(fai), fasta, fai]
+    }
 
     emit:
-    plain_fasta = ch_plain_fasta       // channel: [ val(meta), path/to/fasta ]
-    versions    = ch_versions          // channel: versions.yml
+    fasta_fai = ch_fasta_fai       // channel: [ val(meta), path/to/fasta, path/to/fai ]
+    versions  = ch_versions        // channel: versions.yml
 }
 
+// Read the .fai file to extract the number of sequences, the maximum and total sequence length
+// Inspired from https://github.com/nf-core/rnaseq/blob/3.10.1/lib/WorkflowRnaseq.groovy
```

Comment:
Should this be in `lib` instead of here?

Reply:
I think it's better to keep the function here so that everything is in this one file, which is not very long (110 lines)